### PR TITLE
rustbuild: Avoid some extraneous rustc compiles on cross builds

### PR DESCRIPTION
**Pull request by alexcrichton**
This tweaks a few locations here and there to avoid compiling rustc too many times on our cross-builders on CI.

Closes https://github.com/rust-lang/rust/issues/44132

### CODE REVIEW

File: src/bootstrap/doc.rs [hunk 1]
```diff
@@ -671,7 +671,7 @@ impl Step for ErrorIndex {
 
         builder.ensure(compile::Rustc {
             compiler: builder.compiler(0, build.build),
-            target,
+            target: build.config.build,
         });
```
Comment (Mark-Simulacrum) [inline src/bootstrap/doc.rs]: `ErrorIndex` depends on `Rustc` already so this is useless; let's remove it.

File: src/bootstrap/lib.rs [hunk 2]
```diff
@@ -718,7 +718,7 @@ impl Build {
     fn force_use_stage1(&self, compiler: Compiler, target: Interned<String>) -> bool {
         !self.config.full_bootstrap &&
             compiler.stage >= 2 &&
-            self.hosts.iter().any(|h| *h == target)
+            (self.hosts.iter().any(|h| *h == target) || target == self.build)
```
Comment (Mark-Simulacrum) [inline src/bootstrap/lib.rs]: Hm, something is wrong if this is necessary. I believe we try to make the first host `self.build` in `config.rs`; could you check on that logic?

**Review by Mark-Simulacrum — CHANGES_REQUESTED** [review]

File: src/bootstrap/lib.rs (hunk 2 above)
Comment (alexcrichton) [reply src/bootstrap/lib.rs]: Yeah I tried reordering this in `config.rs` but then it ended up building too much b/c it's using `config.host` to seed the matrix of what to build presumably? That is, when we do `--host foo`, we only want the host binaries for `foo` and not for the build triple. Do you think there's a better way to solve this though?

Comment (Mark-Simulacrum) [reply src/bootstrap/lib.rs]: Probably not then -- I think our current separation between target/host/build isn't really well thought out and probably various build steps build on the wrong platform (i.e., not optimal time-wise) today. I'm hoping to categorize the list of steps and make the distinction clearer.

**Comment by alexcrichton** [comment]
Ok fixed up the first comment, re-r? @Mark-Simulacrum

**Review by Mark-Simulacrum — APPROVED** [review]

**Comment by arielb1** [comment]
I'm too wussy to roll this up, but this should help us get rid of our queue
so
@bors p=1

**Pull request merged**